### ISSUE TITLE
no-param-reassign reports error on destructured argument assignment which would not mutate the arguments object

### ISSUE THREAD

**Issue by hoistedvar** [eslint/eslint#8283]
**Tell us about your environment**

* **ESLint Version:** 3.18
* **Node Version:** 7.7.3
* **npm Version:** 4.1.2

**What parser (default, Babel-ESLint, etc.) are you using?** default

**What did you do? Please include the actual source code causing the issue.**

```js
/* eslint no-param-reassign: "error" */
function test(indexed, { named } = {}) {
	indexed = 'value';
	named = 'value';
	return indexed + named;
}
test('ignored', { named: 'ignored' });
```

**What did you expect to happen?**
Should just report error of `no-param-reassign` for assignment to `indexed` as this is the only parameter to be indexed in the `arguments` object. 

**What actually happened? Please include the actual, raw output from ESLint.**
Reports `no-param-reassign` for both `indexed` and `named`, as named is a destructured parameter it is not indexed in the arguments object and therefore should not trigger this rule.

This rule is to prevent inadvertent modification of the `arguments` object (and optionally preventing mutating parameters via the `props` option) as outlined in the rule description. Assignments which do not inadvertently modify `arguments` should not trigger this error. If the `props` option is used then this should still apply to property modifications of destructured arguments as this will still mutate the incoming parameter value.

**Comment by not-an-aardvark** [comment]
Thanks for the report. However, I think this is working as intended.

You mentioned that "This rule is to prevent inadvertent modification of the `arguments` object", but I don't think that's the main purpose of the rule. For strict mode functions, the `arguments` object wouldn't get modified anyway, so if that were the main purpose, the rule wouldn't need to report strict mode functions:

```js
(function(a) {
  a = 1;
  return arguments[0];
})(0); // => 1

(function(a) {
  'use strict';
  a = 1;
  return arguments[0];
})(0); // => 0
```

I think the main purpose of this rule is to avoid reassignment because it can cause confusion for developers. In other words, variables from the parameter list should be treated as if they were declared with `const`.

**Comment by alberto** [comment]
Closing this issue as it looks like the question has been answered. Please feel free to visit us in the ESLint Gitter if you have any other issues!

**Comment by chrisnicola** [comment]
I just ran into this and have to disagree. For those that want to avoid developer confusing `props: true` is available. If `props: false` is set then I would expect destructured arguments to be "ok" as I don't really distinguish between these uses. 

As it stands now there is no option for people who simply want to prevent modifying `arguments`. At least consider adding an option to support destructured arguments as well as props.
